p2p/listener.go

```diff
@@ -88,17 +92,29 @@ func NewDefaultListener(protocol string, lAddr string, UPNP bool, logger log.Log
 		panic(err)
 	}
 
+	inAddrAny := lAddrIP == "" || lAddrIP == "0.0.0.0"
+
 	// Determine external address...
 	var extAddr *NetAddress
-	if UPNP {
+	if cfg.UPNP {
 		// If the lAddrIP is INADDR_ANY, try UPnP
-		if lAddrIP == "" || lAddrIP == "0.0.0.0" {
+		if inAddrAny {
 			extAddr = getUPNPExternalAddress(lAddrPort, listenerPort, logger)
 		}
 	}
+
+	if cfg.ExternalAddress != "" {
```

Comment:
I think this should be the first check. And only if external_address is "", we should proceed with getUPNPExternalAddress

```
if cfg.ExternalAddress != "" {
  ...
} else if cfg.UPNP && inAddrAny {
  ...
}
```